internal/api/impl/auth/oauth.go

```diff
@@ -0,0 +1,128 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/labstack/echo/v4"
+	"github.com/perses/perses/internal/api/shared"
+	"github.com/perses/perses/internal/api/shared/crypto"
+	"github.com/perses/perses/internal/api/shared/route"
+	"github.com/perses/perses/pkg/model/api/config"
+	"golang.org/x/oauth2"
+)
+
+type oAuthEndpoint struct {
+	conf            *oauth2.Config
+	jwt             crypto.JWT
+	tokenManagement tokenManagement
+	slugID          string
+	userInfoURL     string
+}
+
+func newOAuthEndpoint(params config.OAuthProvider, jwt crypto.JWT) authEndpoint {
+	conf := &oauth2.Config{
+		ClientID:     string(params.ClientID),
+		ClientSecret: string(params.ClientSecret),
+		Scopes:       params.Scopes,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  params.AuthURL,
+			TokenURL: params.TokenURL,
+		},
+		RedirectURL: params.RedirectURI,
+	}
+
+	return &oAuthEndpoint{
+		conf:            conf,
+		jwt:             jwt,
+		tokenManagement: tokenManagement{jwt: jwt},
+		slugID:          params.SlugID,
+		userInfoURL:     params.UserInfosURL,
+	}
+}
+
+func (e *oAuthEndpoint) CollectRoutes(g *route.Group) {
+	g.GET(fmt.Sprintf("/%s/%s", e.String(), loginPath), e.authHandler, true)
+	g.GET(fmt.Sprintf("/%s/%s", e.String(), callbackPath), e.codeExchangeHandler, true)
+}
+
+func (e *oAuthEndpoint) String() string {
+	return strings.Join([]string{"oauth", e.slugID}, "-")
+}
+
+// TODO(cegarcia): Check necessity of verifier + check state?
+func (e *oAuthEndpoint) authHandler(ctx echo.Context) error {
+	verifier := oauth2.GenerateVerifier()
+
+	// Redirect user to consent page to ask for permission
+	// for the scopes specified above.
+	url := e.conf.AuthCodeURL("state", oauth2.AccessTypeOnline, oauth2.S256ChallengeOption(verifier))
+
+	return ctx.Redirect(302, url)
+}
+
+// TODO(cegarcia): Check necessity of verifier + check state?
+// TODO(cegarcia): User Infos retrieval is currently strongly coupled with github provider (GET /user and take login from the json `login` field) but it's probably not the case with others. We should have user infos discovery to try to get the login from different ways (tokens/api with different claim paths/json paths)
+func (e *oAuthEndpoint) codeExchangeHandler(ctx echo.Context) error {
+	code := ctx.QueryParam("code")
+	redirectURI := ctx.QueryParam("redirect_uri")
+	if redirectURI == "" {
+		redirectURI = "/"
+	}
+	verifier := oauth2.GenerateVerifier()
+
+	token, err := e.conf.Exchange(ctx.Request().Context(), code, oauth2.VerifierOption(verifier))
+	if err != nil {
+		return shared.InternalError
+	}
+
+	resp, err := e.conf.Client(ctx.Request().Context(), token).Get(e.userInfoURL)
+	if err != nil {
+		return err
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	var userInfos struct {
```

Comment:
probably we can declare this struct outside of the method. It would avoid to redeclare it everytime this method is called

Reply:
yes I will

Note also that this all "login discovery" needs to be enriched a lot, as it's not necessarily a standard to have login in the user infos payload. I'm pretty sure that we don't support a lot more than Github provider with that... 
I was not sure it's the right way but this is also what Grafana is doing, checking it in several places, the token, user infos answer.. I don't like it that much but I don't have better option. Except having an oauth provider type dedicated by provider, which is not better